### DISCUSSION TITLE
Were the new "special folders" introduced in Vista ever added to System.Environment.SpecialFolder?

### DISCUSSION THREAD

**Discussion by RayKoopa** [dotnet/runtime#68401] [Q&A] [answered]
[`System.Environment.GetFolderPath`](https://docs.microsoft.com/en-us/dotnet/api/system.environment.getfolderpath) accepts an `enum` parameter of type [`System.Environment.SpecialFolder`](https://docs.microsoft.com/en-us/dotnet/api/system.environment.specialfolder). It can be used to retrieve special user folder like "(My) Documents" or "(My) Pictures". However, the `enum` was seemingly never extended to support querying the paths to "newer" user folders introduced in Windows Vista (2007!) and later, like the "(My) Downloads" folder.

I found several (old) discussions on tackling this problem, but effectively, it still does not seem possible to query those folders in .NET 7:
- https://github.com/dotnet/runtime/issues/15419 wants to "complete" the enum. The issue was set to resolved by the msftbot, but the changes linked in the discussion don't actually mention the "new" user folders.
- https://github.com/dotnet/dotnet-api-docs/issues/791 was submitted from Microsoft Docs apparently as a feature request to add this support. The user submitting that issue linked to it in a comment under my [CodeProject article](https://www.codeproject.com/Articles/878605/Getting-all-Special-Folders-in-NET) (which explains the current solution to get these paths via p/invoking [`SHGetKnownFolderPath`](https://docs.microsoft.com/en-us/windows/win32/api/shlobj_core/nf-shlobj_core-shgetknownfolderpath)). Also, a community contributor project to add support for this was mentioned in this issue, but I also cannot find that one.

What's the state of this? I see incorrect "special" folder path determiniation relatively often, and wish this would finally be added to .NET. 
- The seemingly [most commonly queried folder](https://stackoverflow.com/questions/3795023/how-to-programmatically-derive-windows-downloads-folder-userprofile-downloads/71954266) is the user Downloads folder, and it has been introduced in Vista _15 years ago_.
- There are too many ways to determine the path incorrectly. Many people hardcode parts of the path and / or assume it wasn't redirected somewhere outside of the user folder. Many people use a deprecated registry key that was only valid during a Windows 95 Beta, which was kept as more people started to rely on it; and given .NET still doesn't support a "proper" way to retrieve these paths, more will rely on it in the future. Even Raymond Chen [wrote about it (2003)](https://devblogs.microsoft.com/oldnewthing/20031103-00/?p=41973) in [multiple posts (2011)](https://devblogs.microsoft.com/oldnewthing/20110322-00/?p=11163).

Can we finally end the _extraordinarily_ long and sad story about retrieving new user folder paths, or did I just not find the new .NET way to retrieve these? :-)

**Comment by am11** [comment, marked as answer]
There are two API proposals on this topic: #25577 and #554. In the implementation, we can refer to https://docs.microsoft.com/en-us/windows/win32/shell/knownfolderid to find the missing GUIDs in https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/Common/src/Interop/Windows/Shell32/Interop.SHGetKnownFolderPath.cs#L22 `Downloads` can also be supported in Desktop Unix (`$HOME/Downloads` is available in most desktop environments).

**Reply by RayKoopa** [reply]
Thank you for finding some newer discussions on this topic and clarifying the current state of this. I [left a comment](https://github.com/dotnet/runtime/issues/554#issuecomment-1107636092) on one of them for further discussion.
